### PR TITLE
[GAP-25] Golem Certificates structure

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**
resolves: https://github.com/golemfactory/golem-internal/issues/500

### CODE REVIEW

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:12]:
```suggestion
Some Golem features are more sensitive because of potential malicious usage. A good example is outbound network traffic, where malicious user could cause harm if he was able to access any internet resource. In GAP-5 and GAP-4 we decided to limit Requestor abilities by introducing Computational Manifest which defines allowed commands and domains access. Payload Manifest is signed to confirm that image is safe to use.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:14]:
```suggestion
Outbound network is not the only use case which requires certificates. Soon we will have other cases and we need unify certificates usage to allow different sets of permissions.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:18]:
```suggestion
Golem use cases allowing for interaction with external world (like outbound, inbound network traffic) put `Providers` in danger of malicious behavior from Requestor side. To alleviate this danger, in GAP-5 and GAP-4 we decided to limit Requestor abilities by introducing Computational Manifest which should be signed by trusted party.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:20]:
```suggestion
Currently `Providers` are able to choose certificates they trust, by adding them to Provider configuration. This means that any party is able to gain `Providers'` trust, distribute their own certificates and validate safety and security of different Payloads. Despite this, we are aware that gaining trust can be difficult, so Golem Factory has to take initiative to be at least initial source of trust in the Network.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:22]:
```suggestion
Safety and security of Payloads ran on `Providers'` machines can be ensured in few ways:
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:24]:
```suggestion
- Audit of Payload Manifest and VM image, to validate if permissions are low enough for image to be safe
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:28]:
```suggestion
When designing certificates structure that will allow to meet requirements mentioned above, we must keep in mind future use cases that might require signing with certificates:
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:33]:
```suggestion
This means that certificates will be used to sign different kinds of permissions. In case of transferring certification to other companies or signing legal Agreements, we need to ensure that we keep control which permissions they will be able to use.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:35]:
```suggestion
In the future of Golem Network we expect that new, independent features will be implemented by community and they may need to use certification. This document aims to standardize this approach.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:44]:
```suggestion
In further documentation I assume we got number `1.3.6.1.4.1.60000`, until we acquire real OID.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:67]:
```suggestion
To transfer some set of permissions to external entity, Golem Factory has to sign its certificate. This will create certificates chain that will be later validated by Agent application. For signature to be valid, both certificate and its parent has to contain specific allowance (or `all`).
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:78]:
```suggestion
then Payload Manifest signature is valid and `Parter's certificate` is allowed to sign Manifests, but has no rights to sign inbound related features.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:89]:
```suggestion
then `Partner's certificate` is not allowed to sign manifests and Agent application should reject such Proposals.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:90]:
```suggestion
Note that the certificate chain is incorrect, because `Partner's certificate` shouldn't be signed by `Golem Intermediate` in the first place.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:64]: How do we manage/maintain the list of allowed `GolemCertificatesAllowance` values? This calls for a public, 'official' list, maybe maintained as part of Golem Standards (here: https://github.com/golemfactory/golem-architecture/tree/master/standards ?)

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:177]:
```suggestion
Root certificate is always self-signed certificate. We could sign it using [Golem Multisig](https://etherscan.io/address/0x7da82c7ab4771ff031b66538d2fb9b0b047f6cf9) contract to allow anyone validate its authenticity.
```

Comment (stranger80) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:177]: I strongly recommend leveraging a well-established public CA as issuer of Golem Root (or in this case - Golem Intermediate) certificates.  
We do not want to be a public CA in full sense, and we don't want to be legally liable for faults in managing the Root CA certs. Buying a Root/Intermediate cert from a public CA delegates risk away from GolemFactory, and will make Golem certificates be more accepted by the users in the "industry" (esp. the enterprise users). 
A concise article to back the above comment can be found here: https://www.thesslstore.com/blog/how-to-become-a-certificate-authority/

**Review by stranger80 — CHANGES_REQUESTED** [review]
A very well written GAP! I appreciate thorough descriptions and justifications of the concepts proposed. 

I have added a number of punctuation edits, plus an imho significant comment on Root certs.

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (pwalski) [reply gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:177]: I agree it is better to delegate reponsibility to some Public CA but it creates a question whether we should not delegate responsibility of keeping Public CA root certificates up to date to user/OS and to just take them from OS? It will also make easier for an app author to distribute their intermediate certificate signed by Public CA because there will be no necessity to import (and prepare) a whole chain.
If we will stick to internal certificate storage using Public CA will make having a proper certificate revocation handling in Provider's keystore an absolute must (there is already a task for it).

Comment (pwalski) [reply gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:177]: On second thought I see a problem.
In some way we are turning x509 certificates into an authorization method.
The way we wanted to allow providers to control whom to trust or not was by importing App author's public certificates into provider's keystore. Right now we do not verify whether imported certificate (or chain) meets any criteria, we just check if it is valid and it gets saved inside a keystore. The moment Public CA certs will end up in Provider's keystore any Demand with manifest signed directly by Public CA certificate will get accepted. Anyone can aquire certificate signed by same Public CA so suddenly Provider starts to trust everyone.
Verifying `GolemCertificatesAllowance` in a certificate incoming with Demand will not help. What would need to be verified will be `GolemCertificatesAllowance` of the certificate signing the certificate incoming in a Demand (lets say `app-author` allowance). Alternatively we could just check if the fields like company name are the same between the certificate incoming in Demand and the certificate signing incoming one.

Comment (nieznanysprawiciel) [reply gaps/gap-25_golem_certificates/gap-25_golem_certificates.md]: @stranger80 After reading your link, I think, I imagined our role rather as private CA not public CA.
I wanted self-signed certificate, because I wasn't not sure if any CA authority will sign our certificate, if we are adding our custom extensions of unknown (from their perspective) purpose.

On the other side: does having self-signed certificate make us immediately public CA?

And my other concern is: although we could be signed by public CA, we can't have our parent certificate on Provider, because it won't have any `GolemCertificatesAllowance`, so the whole chain would be rejected by Provider.
This relates to your comment either @pwalski

Comment (nieznanysprawiciel) [reply gaps/gap-25_golem_certificates/gap-25_golem_certificates.md]: The second thing: I talked with @prekucki and we can't sign our certificate using Multisig anyway, because signature comes from only one holder and contract logic ensures, that all holders agreed to sign.

Comment (nieznanysprawiciel) [reply gaps/gap-25_golem_certificates/gap-25_golem_certificates.md]: And the last thing: I'm wondering if we shouldn't use blockchain for storing/handling these certificates.

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (nieznanysprawiciel) [reply gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:64]: I wrote somewhere, that we will update list in this GAP, but maybe it is good idea to maintain it in standards. But we must take care that we have links to standards everywhere, because sometimes I was having hard time to find something.

**Comment by shadeofblue** [comment]
@evik42 @nieznanysprawiciel I'll wait until the other pull request is merged with my review as it's hard to read this one knowing an update is due

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (shadeofblue) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:12]: I'd use a gender-neutral term ;)

```suggestion
Some Golem features are more sensitive because of potential malicious usage. A good example is outbound network traffic, where malicious user could cause harm if they were able to access any internet resource. In GAP-5 and GAP-4 we decided to limit Requestor abilities by introducing Computational Manifest which defines allowed commands and domains access. Payload Manifest is signed to confirm that image is safe to use.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (shadeofblue) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:18]:
```suggestion
Golem use cases allowing for interaction with external world (like outbound, inbound network traffic) put `Providers` in danger of malicious behavior from Requestor side. To alleviate this danger, in GAP-5 and GAP-4 we decided to limit Requestor abilities by introducing Computational Manifest which should be signed by a trusted party.
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (shadeofblue) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:22]:
```suggestion
Safety and security of Payloads run on `Providers'` machines can be ensured in few ways:
```

File: gaps/gap-25_golem_certificates/gap-25_golem_certificates.md
Comment (shadeofblue) [inline gaps/gap-25_golem_certificates/gap-25_golem_certificates.md:24]:
```suggestion
- Audit of the Payload Manifest and of the VM image, to validate if permissions are low enough for image to be safe
```

**Comment by evik42** [comment]
This is out of date and will be closed in favor of #67 when that is accepted.

**Pull request closed without merging**
